### PR TITLE
feat(dashboard): introduced password expiration for the Dashboard

### PR DESCRIPTION
**Pull request by lafirest**
Fixes EMQX-13506

Release version: v/e5.9.0

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [x] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx_dashboard/src/emqx_dashboard_admin.erl [hunk 1]
```diff
@@ -280,6 +288,21 @@ verify_hash(Origin, SaltHash) ->
 sha256(SaltBin, Password) ->
     crypto:hash('sha256', <<SaltBin/binary, Password/binary>>).
 
+verify_password_expiration(0, _User) ->
+    {ok, #{}};
+verify_password_expiration(ExpiredTime, #?ADMIN{extra = #{password_ts := Ts}}) ->
```
Comment (zmstone) [inline apps/emqx_dashboard/src/emqx_dashboard_admin.erl:293]:
```suggestion
verify_password_expiration(ValidityDuration, #?ADMIN{extra = #{password_ts := Ts}}) ->
```

File: apps/emqx_dashboard/src/emqx_dashboard_admin.erl [hunk 2]
```diff
@@ -280,6 +288,21 @@ verify_hash(Origin, SaltHash) ->
...
+    Diff = Ts + ExpiredTime - erlang:system_time(second),
+    {ok, #{password_expire_in_seconds => Diff}};
+verify_password_expiration(ExpiredTime, #?ADMIN{extra = Extra} = User) ->
+    case
+        unsafe_update_user(User#?ADMIN{extra = Extra#{password_ts => erlang:system_time(second)}})
+    of
+        ok ->
+            {ok, #{password_ts_init => true, password_expire_in_seconds => ExpiredTime}};
```
Comment (zmstone) [inline apps/emqx_dashboard/src/emqx_dashboard_admin.erl:301]: question, why is `password_ts_init` needed ?

File: apps/emqx_dashboard/src/emqx_dashboard_admin.erl [hunk 3]
```diff
@@ -378,9 +409,13 @@ check(Username, Password) ->
...
-            emqx_dashboard_token:sign(User, Password);
+    ExpiredTime = emqx:get_config([dashboard, password_expired_time], 0),
+    maybe
+        {ok, User} ?= check(Username, Password),
+        {ok, Result} ?= verify_password_expiration(ExpiredTime, User),
+        {ok, Role, Token} ?= emqx_dashboard_token:sign(User, Password),
+        {ok, Result#{role => Role, token => Token}}
+    else
         Error ->
             Error
```
Comment (zmstone) [inline apps/emqx_dashboard/src/emqx_dashboard_admin.erl:420]: we can remove this `else` clause ?

File: rel/i18n/emqx_dashboard_api.hocon [hunk 4]
```diff
@@ -87,4 +87,10 @@ role.desc:
 backend.desc:
 """User account source"""
 
+password_ts_init.desc:
+"""From version 5.9, EMQX introduced a password expiration time, and old accounts are automatically updated for compatibility. This field marks that this account has been updated in this login."""
+
+password_expire_in_seconds.desc:
+"""Password remaining expiration time."""
```
Comment (zmstone) [inline rel/i18n/emqx_dashboard_api.hocon:94]:
```suggestion
"""Time in seconds until the password expires."""
```

**Review by HJianBo — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-6680)

**Pull request merged**